File: app/src/main/java/dev/ragnarok/fenrir/fragment/conversation/ConversationPhotosFragment.java
package dev.ragnarok.fenrir.fragment.conversation;

import static android.app.Activity.RESULT_OK;

import android.content.Intent;
import android.os.Bundle;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.Extra;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.adapter.fave.FavePhotosAdapter;
import dev.ragnarok.fenrir.model.Photo;
import dev.ragnarok.fenrir.model.TmpSource;
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory;
import dev.ragnarok.fenrir.mvp.presenter.conversations.ChatAttachmentPhotoPresenter;
import dev.ragnarok.fenrir.mvp.view.conversations.IChatAttachmentPhotosView;
import dev.ragnarok.fenrir.place.PlaceFactory;

public class ConversationPhotosFragment extends AbsChatAttachmentsFragment<Photo, ChatAttachmentPhotoPresenter,
        IChatAttachmentPhotosView> implements FavePhotosAdapter.PhotoSelectionListener, FavePhotosAdapter.PhotoConversationListener, IChatAttachmentPhotosView {

    private final ActivityResultLauncher<Intent> requestPhotoUpdate = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == RESULT_OK && result.getData() != null && result.getData().getExtras() != null) {
                    int ps = result.getData().getExtras().getInt(Extra.POSITION);
                    ((FavePhotosAdapter) getAdapter()).updateCurrentPosition(ps);
                    mRecyclerView.scrollToPosition(ps);
                }
            });

    @Override
    protected RecyclerView.LayoutManager createLayoutManager() {
        int columns = getResources().getInteger(R.integer.photos_column_count);
        return new GridLayoutManager(requireActivity(), columns);
    }

    @Override
    public RecyclerView.Adapter<?> createAdapter() {
        FavePhotosAdapter apiPhotoFavePhotosAdapter = new FavePhotosAdapter(requireActivity(), Collections.emptyList());
        apiPhotoFavePhotosAdapter.setPhotoSelectionListener(this);
        apiPhotoFavePhotosAdapter.setPhotoConversationListener(this);
        return apiPhotoFavePhotosAdapter;
    }

    @Override
    public void onPhotoClicked(int position, Photo photo) {
        callPresenter(p -> p.firePhotoClick(position, photo));
    }

    @NonNull
    @Override
    public IPresenterFactory<ChatAttachmentPhotoPresenter> getPresenterFactory(@Nullable Bundle saveInstanceState) {
        return () -> {
            int accountId = requireArguments().getInt(Extra.ACCOUNT_ID);
            int peerId = requireArguments().getInt(Extra.PEER_ID);
            return new ChatAttachmentPhotoPresenter(peerId, accountId, saveInstanceState);
        };
    }

    @Override
    public void displayAttachments(List<Photo> data) {
        FavePhotosAdapter adapter = (FavePhotosAdapter) getAdapter();
        adapter.setData(data);
    }

    @Override
    public void goToTempPhotosGallery(int accountId, @NonNull TmpSource source, int index) {
        PlaceFactory.getTmpSourceGalleryPlace(accountId, source, index).setActivityResultLauncher(requestPhotoUpdate).tryOpenWith(requireActivity());
    }

    @Override
    public void goToTempPhotosGallery(int accountId, long ptr, int index) {
        PlaceFactory.getTmpSourceGalleryPlace(accountId, ptr, index).setActivityResultLauncher(requestPhotoUpdate).tryOpenWith(requireActivity());
    }

    @Override
    public void onGoPhotoConversation(@NonNull Photo photo) {
        callPresenter(p -> p.fireGoToMessagesLookup(photo.getMsgPeerId(), photo.getMsgId()));
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/conversation/ConversationPhotosFragment.kt
package dev.ragnarok.fenrir.fragment.conversation

import android.app.Activity
import android.os.Bundle
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dev.ragnarok.fenrir.Extra
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.adapter.fave.FavePhotosAdapter
import dev.ragnarok.fenrir.adapter.fave.FavePhotosAdapter.PhotoConversationListener
import dev.ragnarok.fenrir.model.Photo
import dev.ragnarok.fenrir.model.TmpSource
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory
import dev.ragnarok.fenrir.mvp.presenter.conversations.ChatAttachmentPhotoPresenter
import dev.ragnarok.fenrir.mvp.view.conversations.IChatAttachmentPhotosView
import dev.ragnarok.fenrir.place.PlaceFactory.getTmpSourceGalleryPlace

class ConversationPhotosFragment :
    AbsChatAttachmentsFragment<Photo, ChatAttachmentPhotoPresenter, IChatAttachmentPhotosView>(),
    FavePhotosAdapter.PhotoSelectionListener, PhotoConversationListener, IChatAttachmentPhotosView {
    private val requestPhotoUpdate = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK && result.data != null && (result.data
                ?: return@registerForActivityResult)
                .extras != null
        ) {
            val ps = ((result.data ?: return@registerForActivityResult).extras
                ?: return@registerForActivityResult).getInt(Extra.POSITION)
            (adapter as FavePhotosAdapter?)?.updateCurrentPosition(ps)
            mRecyclerView?.scrollToPosition(ps)
        }
    }

    override fun createLayoutManager(): RecyclerView.LayoutManager {
        val columns = resources.getInteger(R.integer.photos_column_count)
        return GridLayoutManager(requireActivity(), columns)
    }

    override fun createAdapter(): RecyclerView.Adapter<*> {
        val apiPhotoFavePhotosAdapter = FavePhotosAdapter(requireActivity(), emptyList())
        apiPhotoFavePhotosAdapter.setPhotoSelectionListener(this)
        apiPhotoFavePhotosAdapter.setPhotoConversationListener(this)
        return apiPhotoFavePhotosAdapter
    }

    override fun onPhotoClicked(position: Int, photo: Photo) {
        presenter?.firePhotoClick(
            position
        )
    }

    override fun getPresenterFactory(saveInstanceState: Bundle?): IPresenterFactory<ChatAttachmentPhotoPresenter> {
        return object : IPresenterFactory<ChatAttachmentPhotoPresenter> {
            override fun create(): ChatAttachmentPhotoPresenter {
                val accountId = requireArguments().getInt(Extra.ACCOUNT_ID)
                val peerId = requireArguments().getInt(Extra.PEER_ID)
                return ChatAttachmentPhotoPresenter(peerId, accountId, saveInstanceState)
            }
        }
    }

    override fun displayAttachments(data: MutableList<Photo>) {
        val adapter = adapter as FavePhotosAdapter?
        adapter?.setData(data)
    }

    override fun goToTempPhotosGallery(accountId: Int, source: TmpSource, index: Int) {
        getTmpSourceGalleryPlace(accountId, source, index).setActivityResultLauncher(
            requestPhotoUpdate
        ).tryOpenWith(requireActivity())
    }

    override fun goToTempPhotosGallery(accountId: Int, ptr: Long, index: Int) {
        getTmpSourceGalleryPlace(
            accountId,
            ptr,
            index
        ).setActivityResultLauncher(requestPhotoUpdate).tryOpenWith(requireActivity())
    }

    override fun onGoPhotoConversation(photo: Photo) {
        presenter?.fireGoToMessagesLookup(
            photo.msgPeerId,
            photo.msgId
        )
    }
}